core/sail/extensible-store/src/main/java/org/eclipse/rdf4j/sail/extensiblestore/evaluationstatistics/ExtensibleDynamicEvaluationStatistics.java

```diff
@@ -0,0 +1,337 @@
+
+			min = Math.min(min, getSubjectCardinality(sp.getSubjectVar()));
+			min = Math.min(min, getPredicateCardinality(sp.getPredicateVar()));
+			min = Math.min(min, getObjectCardinality(sp.getObjectVar()));
+
+			if (sp.getSubjectVar().getValue() != null && sp.getPredicateVar().getValue() != null) {
+				min = Math.min(min,
+						getHllCardinality(subjectPredicateIndex, subjectPredicateIndex_removed,
+								sp.getSubjectVar().getValue(),
+								sp.getPredicateVar().getValue()));
+			}
+
+			if (sp.getPredicateVar().getValue() != null && sp.getObjectVar().getValue() != null) {
+				min = Math.min(min,
+						getHllCardinality(predicateObjectIndex, predicateObjectIndex_removed,
+								sp.getPredicateVar().getValue(),
+								sp.getObjectVar().getValue()));
+			}
+
+			return min;
+
+		}
+
+		@Override
+		protected double getSubjectCardinality(Var var) {
+			if (var.getValue() == null) {
+				return size.estimateCardinality();
+			} else {
+				return getHllCardinality(subjectIndex, subjectIndex_removed, var.getValue());
+			}
+
+		}
+
+		@Override
+		protected double getPredicateCardinality(Var var) {
+			if (var.getValue() == null) {
+				return size.estimateCardinality();
+			} else {
+				return getHllCardinality(predicateIndex, predicateIndex_removed, var.getValue());
+			}
+		}
+
+		@Override
+		protected double getObjectCardinality(Var var) {
+			if (var.getValue() == null) {
+				return size.estimateCardinality();
+			} else {
+				return getHllCardinality(objectIndex, objectIndex_removed, var.getValue());
+			}
+		}
+
+		@Override
+		protected double getContextCardinality(Var var) {
+			if (var.getValue() == null) {
+				return defaultContext.estimateCardinality() - defaultContext_removed.estimateCardinality();
+			} else {
+				return getHllCardinality(contextIndex, contextIndex_removed, var.getValue());
+			}
+		}
+	};
+
+	private double getHllCardinality(HyperLogLogCollector[][] index, HyperLogLogCollector[][] index_removed,
+			Value value1, Value value2) {
+		return index[Math.abs(value1.hashCode() % index.length)][Math.abs(value2.hashCode() % index.length)]
+				.estimateCardinality()
+				- index_removed[Math.abs(value1.hashCode() % index_removed.length)][Math
+						.abs(value2.hashCode() % index_removed.length)]
+								.estimateCardinality();
+	}
+
+	private double getHllCardinality(Map<Integer, HyperLogLogCollector> index,
+			Map<Integer, HyperLogLogCollector> index_removed, Value value) {
+
+		return index.getOrDefault(Math.abs(value.hashCode() % ONE_DIMENSION_INDEX_SIZE), EMPTY_HLL)
+				.estimateCardinality()
+				- index_removed.getOrDefault(Math.abs(value.hashCode() % ONE_DIMENSION_INDEX_SIZE), EMPTY_HLL)
+						.estimateCardinality();
+	}
+
+	@Override
+	public void add(Statement statement, boolean inferred) {
+
+		queue.add(new StatemetQueueItem(statement, inferred, true));
+
+		int size = queueSize.incrementAndGet();
+		if (size > QUEUE_LIMIT && queueThread == null) {
+			startQueueThread();
```

Comment:
What is the role of this thread? It updates the various indexes I gather, is that such an expensive operation that it should be handled asynchronously?

Reply:
It's to do async stats updates. The thread is started if the size of the queue is likely to be above the limit, then runs until the queue is empty. Since the stats are only estimates there is no guarantee for timeliness. Those guarantees are needed when running tests and benchmarking, so there is a poor mans latch that waits for the thread to finish.